src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs

```diff
@@ -11,65 +12,107 @@ internal static class RestoreSqlBackupImpl
     {
         // if database name is not provided, dbname from the backup is used.
         // if newStoragePath is not provided, system defaults are used
-        internal static void RestoreSqlBackup(ICakeContext context, String connectionString, FilePath backupFile, RestoreSqlBackupSettings settings)
+        internal static void RestoreSqlBackup(ICakeContext context, String connectionString, RestoreSqlBackupSettings settings, IList<FilePath> backupFiles, IList<FilePath> differentialBackupFiles = null)
         {
             using (var connection = SqlServerAliasesImpl.OpenSqlConnection(context, connectionString))
             {
-                var oldDbName = GetDatabaseName(backupFile, connection);
-                var newDatabaseName = settings.NewDatabaseName ?? oldDbName;
-                context.Log.Information($"Using database name '{newDatabaseName}' to be a name for the restored database");
+                bool hasDifferentialBackup = differentialBackupFiles != null && differentialBackupFiles.Any();
+                var mainCommand = GetSqlCommand(context, connection, settings, backupFiles, settings.BackupSetFile, hasDifferentialBackup);
+                mainCommand.ExecuteNonQuery();
+                if (hasDifferentialBackup)
+                {
+                    var differentialCommand = GetSqlCommand(context, connection, settings, differentialBackupFiles, settings.DifferentialBackupSetFile, false, false);
+                    differentialCommand.ExecuteNonQuery();
+                }
+            }
+        }
 
-                var logicalNames = GetLogicalNames(backupFile, connection);
+        private static SqlCommand GetSqlCommand(ICakeContext context, SqlConnection connection, RestoreSqlBackupSettings settings, IList<FilePath> backupFiles, int backupSetFile = 1, bool noRecovery = false, bool allowAlterDb = true)
```

Comment:
Given amount of parameters, optional parameters and if-statements, I strongly feel that this method should be split:
`CommandForSingleUserMode`
`CommandForFullBackups`
`CommandForDiffBackups`
`CommandForRestoringMultiUserMode`

Otherwise we are doing too much in one place and that is hard to follow and prone to bugs. Also full/diff backups do not always have the same options available, so these are really different SQL statements.

I'm sure we'll be able to extract methods to populate file names and logical names parameter values so these can be reused for Full/Diff backups. 

After that is done, I'm sure all the optional parameters will be gone and amount of if-statements will be reduced.

Reply:
Ok, I'll do an update. Are you ok to break compatibility to older versions e.g. regarding method signatures, behaviors and/or RestoreSqlBackupSettings?

Or was the above suggestion mostly meant for the "internal" design?

Reply:
I'd rather not break the compatibility. What I meant is for the internal implemenatation.

Reply:
> 
> 
> I'd rather not break the compatibility. What I meant is for the internal implemenatation.

All right. I tried to improve the design a bit.